### PR TITLE
[3] Handle Setting Entity History in IO

### PR DESCRIPTION
**Pull request by JonoPrest**
- Moves Entity History to its own module
- Adds SET | DELETE enum for entity history rows
- Adds custom schema for mapping an entity with nullable fields
- Wires up the new entity history to IO

NOTE: the rollback tests won't pass until I've added the next step of diffing these tables

### CODE REVIEW

File: codegenerator/cli/templates/dynamic/codegen/src/db/Entities.res.hbs [hunk 1]
```diff
@@ -23,257 +24,6 @@ let isIndex = true
 @genType
 type whereOperations<'entity, 'fieldType> = {eq: 'fieldType => promise<array<'entity>>}
 
-module EntityHistory = {
-  type historyFieldsGeneral<'a> = {
```
Comment (JonoPrest) [inline codegenerator/cli/templates/dynamic/codegen/src/db/Entities.res.hbs:27]: Moved to its own module, there are also a number of changes so might be nicer to review diffs of commits after the move.

File: codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+  block_timestamp: 'a,
+  block_number: 'a,
+  log_index: 'a,
+}
+
+type historyFields = historyFieldsGeneral<int>
+
+type entityIdOnly = {id: string}
+let entityIdOnlySchema = S.schema(s => {id: s.matches(S.string)})
+type entityData<'entity> = Delete(entityIdOnly) | Set('entity)
```
Comment (JonoPrest) [inline codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res:14]: Entity data can now be delete or set. Delete still needs the ID, and the only way we can distingish between an entity that already has all nullable fields and a deleted entity is with an additional flag (which is added to the table.)

File: codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+    })
+    nulldict
+  | _ =>
+    Js.Exn.raiseError("Failed creating entityWithNullFields. Expected an object schema for entity")
+  }
+
+  //Gets an entity object with all fields set to null except for the id field
+  let getEntityWithNullFields = (entityId: string) => {
+    entityWithNullFields->Utils.Dict.updateImmutable("id", entityId->Utils.magic)
+  }
```
Comment (JonoPrest) [inline codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res:55]: Added this function that gets an entity with all its fields as nullable except id

File: codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+    block_number: None,
+    log_index: None,
+  }
+
+  S.object(s => {
+    {
+      "current": s.flatten(currentHistoryFieldsSchema),
+      "previous": s.flatten(previousHistoryFieldsSchema),
+      "entityData": s.flatten(nullableEntitySchema),
+      "action": s.field("action", Enums.EntityHistoryRowAction.enum.schema),
```
Comment (JonoPrest) [inline codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res:92]: add a field called "action"

File: codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+          block_timestamp,
+          block_number,
+          log_index,
+        })
+      | {chain_id: None, block_timestamp: None, block_number: None, log_index: None} => None
+      | _ => s.fail("Unexpected mix of null and non-null values in previous history fields")
+      },
+      entityData: switch v["action"] {
+      | SET => v["entityData"]->S.parseAnyOrRaiseWith(entitySchema)->Set
+      | DELETE => v["entityData"]->S.parseAnyOrRaiseWith(entityIdOnlySchema)->Delete
```
Comment (JonoPrest) [inline codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res:115]: Based on the action parse or raise with the full schema or id only schema

File: codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res (hunk 3 above)
Comment (DZakh) [reply codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res:55]: Is it for rescript-schema to pass validation? I think postgress should be fine to have the null fields omitted

File: codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+      },
+    },
+    serializer: v => {
+      let (entityData, action) = switch v.entityData {
+      | Set(entityData) => (
+          entityData->(Utils.magic: 'entity => Js.Dict.t<unknown>),
+          Enums.EntityHistoryRowAction.SET,
+        )
+      | Delete({id}) => (getEntityWithNullFields(id), DELETE)
```
Comment (DZakh) [inline codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res:124]: I don't really like that it'll create an array to return the tuple from the switch, but this is not super critical

File: codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+
+let insertRow = (self: t<'entity>, ~sql, ~historyRow: historyRow<'entity>) => {
+  let row = historyRow->S.serializeOrRaiseWith(self.schema)
+  self.insertFn(sql, row)
+}
+
+let batchInsertRows = (self: t<'entity>, ~sql, ~rows: array<historyRow<'entity>>) => {
+  Utils.Array.awaitEach(rows, async row => {
+    let row = row->S.serializeOrRaiseWith(self.schema)
+    await self.insertFn(sql, row)
```
Comment (DZakh) [inline codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res:156]: Should we maybe have an insertFn which works with multiple items?

Comment (DZakh) [reply codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res:156]: I assume it's going to be faster

**Review by DZakh — APPROVED** [review]
Like it

File: codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res (hunk 6 above)
Comment (JonoPrest) [reply codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res:124]: I can inline it and just have 2 switch statements. Not sure if array allocation or 2 switches is worse

File: codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res (hunk 7 above)
Comment (JonoPrest) [reply codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res:156]: As in - a postgres function that iterates through a set of rows and calls the postgres insert function?

We can try that, might be a bit faster 👍🏼

Think I will just try get everything end to end first and note this as an optimisation possibility.

File: codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res (hunk 3 above)
Comment (JonoPrest) [reply codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res:55]: Hmmm 🤔 let me test it.

File: codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res (hunk 6 above)
Comment (JonoPrest) [reply codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res:124]: Just a quick benchmark for this and they look equivalent to me: 

[rescript code](https://rescript-lang.org/try?version=v11.1.4&code=C4TwDgpgBAhgxsAlgewHZQLxQIJQD5QBCAUMaJFAG4wA2mUAkqsABQDeqArgLYBcUiZgF8AlPigBlYACd2AZxn8F0wQHNRZcNEHAI01LVIQAHrv20o3GKsRx+AchiYAfFHsAjegCIApIgAmEMyIoF6k5NDSyADu9GzEULAIKKj88EhoADQJVLT8OnoGNNlCpDQQwFDI7gBWACrIAEoxuFgs1DT8HSL8UbEYrvGJ5ZUs6SmZuTRiWHLRIXAAFlNQQ+JMrBw8+aIuUCzYk4gAtM5WNnAiOQRSsmzKSrsD+4STcqfntleJpTlrieMsjlEh1sj9iL8RlVag1mtFCPR2nkpj0oH09mtAakoHMFssOqtroxmCwAPozVzYIm3MkUog5IRgqZKebAJYrNYEDbsLh8ARPVwnM7WWzUmTyRQ4gU4j4iuAMkplCrQmobejcrZ8ACMGih1Rqt3oNPukq8ADNkMgvBogA)

[benchmark](https://jsbenchmark.com/#eyJjYXNlcyI6W3siaWQiOiJWdEx6SzlYNG9fejF0WXQwSG43VWIiLCJjb2RlIjoiREFUQS5vYmpUb1Jvd0EoREFUQS5vYmpJbnQpIiwiZGVwZW5kZW5jaWVzIjpbXX0seyJpZCI6Ii1xTDJleHRvYlBMM2xHSUw2NVZCdiIsImNvZGUiOiJEQVRBLm9ialRvUm93QihEQVRBLm9iakludCkiLCJkZXBlbmRlbmNpZXMiOltdfSx7ImlkIjoiNF9taGZ4SkVMZUpNR0ZvQnc1dU5qIiwiY29kZSI6IkRBVEEub2JqVG9Sb3dBKERBVEEub2JqU3RyKSIsImRlcGVuZGVuY2llcyI6W119LHsiaWQiOiJXRGdvS1ItMmRydkVNV3dycmtEajkiLCJjb2RlIjoiREFUQS5vYmpUb1Jvd0IoREFUQS5vYmpTdHIpIiwiZGVwZW5kZW5jaWVzIjpbXX1dLCJjb25maWciOnsibmFtZSI6IkJhc2ljIGV4YW1wbGUiLCJwYXJhbGxlbCI6dHJ1ZSwiZ2xvYmFsVGVzdENvbmZpZyI6eyJkZXBlbmRlbmNpZXMiOltdfSwiZGF0YUNvZGUiOiJmdW5jdGlvbiBvYmpUb1Jvd0EodmFsKSB7XG4gIHZhciBtYXRjaDtcbiAgbWF0Y2ggPSB2YWwuVEFHID09PSBcIkludFwiID8gW1xuICAgICAgXCJBXCIsXG4gICAgICB2YWwubnVtXG4gICAgXSA6IFtcbiAgICAgIFwiQlwiLFxuICAgICAgdmFsLnN0clxuICAgIF07XG4gIHJldHVybiB7XG4gICAgICAgICAgYWN0aW9uOiBtYXRjaFswXSxcbiAgICAgICAgICB2YWw6IG1hdGNoWzFdXG4gICAgICAgIH07XG59XG5cbmZ1bmN0aW9uIG9ialRvUm93Qih2YWwpIHtcbiAgdmFyIHRtcDtcbiAgdG1wID0gdmFsLlRBRyA9PT0gXCJJbnRcIiA_IFwiQVwiIDogXCJCXCI7XG4gIHZhciB0bXAkMTtcbiAgdG1wJDEgPSB2YWwuVEFHID09PSBcIkludFwiID8gdmFsLm51bSA6IHZhbC5zdHI7XG4gIHJldHVybiB7XG4gICAgICAgICAgYWN0aW9uOiB0bXAsXG4gICAgICAgICAgdmFsOiB0bXAkMVxuICAgICAgICB9O1xufVxuXG52YXIgb2JqSW50ID0ge1xuICBUQUc6IFwiSW50XCIsXG4gIG51bTogMVxufTtcblxudmFyIG9ialN0ciA9IHtcbiAgVEFHOiBcIlN0clwiLFxuICBzdHI6IFwiZm9vXCJcbn07XG5cbnJldHVybiB7XG4gIG9ialRvUm93QSAsXG4gIG9ialRvUm93QiAsXG4gIG9iakludCAsXG4gIG9ialN0ciAsXG59In19)

Comment (DZakh) [reply codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res:124]: Ok, nice, I won't care about cases like this anymore

**Pull request merged**